src/rust/engine/graph/README.md

```diff
@@ -0,0 +1,41 @@
+
+At any point (perhaps concurrently with `Node`s running), external threads may call [Graph::invalidate_from_roots](https://github.com/pantsbuild/pants/blob/01372719f7e6f2a3aa0b9f3ce6909991388071ca/src/rust/engine/graph/src/lib.rs#L401-L407) to indicate that the result values of particular `Node` instances are no longer valid for some reason. Because the identity of a `Node` is immutable (see the "Inputs and Identity" section above), invalidation almost always needs to occur because some information that the `Node` fetched using the `Context` (or direct use of syscalls, etc) is no longer valid. In production usage, invalidation occurs due to filesystem changes: a file watching thread [calls invalidate_from_roots](https://github.com/pantsbuild/pants/blob/01372719f7e6f2a3aa0b9f3ce6909991388071ca/src/rust/engine/src/context.rs#L351-L362) for `Node` types that are associated with syscalls to read files when it suspects that files on disk have changed.
+
+Confusingly, "root" in the context of `invalidate_from_roots` means a root in the dependee graph: calling the method for a root `Node` will clear the value of that `Node`, and mark the transitive dependees of that `Node` "dirty". The cleared `Node` will definitely re-run when it is next requested (it goes back to the `NotStarted` state), but dirty `Node`s will only re-run if they cannot be "cleaned": see below.
+
+### Dirtying and Cleaning
+
+When the `Entry` for a `Node` is marked dirty by "Invalidation" (see above) we only _suspect_ that it might need to re-run, because one of its dependencies has been cleared. To determine whether it _actually_ needs to re-run, the next time the dirty `Node` is requested, `Graph::get` (via [`Entry::get_node_result`](https://github.com/pantsbuild/pants/blob/01372719f7e6f2a3aa0b9f3ce6909991388071ca/src/rust/engine/graph/src/entry.rs#L361-L374)) will compare the [`Generation` values](https://github.com/pantsbuild/pants/blob/01372719f7e6f2a3aa0b9f3ce6909991388071ca/src/rust/engine/graph/src/entry.rs#L37-L57) of the dependencies used to compute the previous value of the `Node` with their current `Generation` values. If none of the (`Generation` values) of the `Node` have changed, then it does not need to re-run: this is called "cleaning".
+
+When a `Node` _does_ need to re-run for some reason (either due to having been cleared by invalidation, or being marked dirty and then failing to be cleaned), its previous and new result are compared to determine whether its `Generation` value should increment. The `Generation` value avoids the need to keep multiple copies of a `Node`'s result value in memory over time, and makes cleaning cheaper by converting potentially expensive equality checks into integer comparisons.
+
+## Uncacheable Nodes
+
+TODO!
```

Comment:
Will fill this in, and reference a new ticket about only cleaning `UncachableDependencies` nodes once per session.